src/cpp/client.cpp

```diff
@@ -1169,6 +1396,167 @@ inline CommandReply Client::_get_dataset_metadata(const std::string& name)
+inline void Client::_get_and_add_dataset_tensor(DataSet& dataset,
+                                                const std::string& name,
+                                                const std::string& key)
+{
+    // Run tensor retrieval command
+    CommandReply reply = _redis_server->get_tensor(key);
+
+    // Extract tensor properties from command reply
+    std::vector<size_t> reply_dims = GetTensorCommand::get_dims(reply);
+    std::string_view blob = GetTensorCommand::get_data_blob(reply);
+    SRTensorType type = GetTensorCommand::get_data_type(reply);
+
+    // Add tensor to the dataset
+    dataset._add_to_tensorpack(name, (void*)blob.data(), reply_dims,
+                               type, SRMemLayoutContiguous);
+}
+
+inline std::vector<DataSet>
+Client::_get_dataset_list_range(const std::string& list_name,
+                                const int start_index,
+                                const int end_index)
+{
+    // Build the list key
+    std::string list_key = _build_list_key(list_name, true);
+
+    // Build the command to retrieve the list
+    SingleKeyCommand cmd;
+    cmd << "LRANGE" << Keyfield(list_key);
+    cmd << std::to_string(start_index);
+    cmd << std::to_string(end_index);
+
+    // Run the command to retrive the list
+    CommandReply reply = _run(cmd);
+
+    // Check for reply errors and correct type
+    if (reply.has_error() > 0)
+        throw SRRuntimeException("GET command failed. The aggregation "\
+                                 "list could not be retrieved.");
+
+    if (reply.redis_reply_type() != "REDIS_REPLY_ARRAY")
+        throw SRRuntimeException("An unexpected type was returned for "
+                                 "for the aggregation list.");
+
+    // Create CommandList for retrieving all metadata values in pipeline
+    CommandList metadata_cmd_list;
+
+    for (size_t i = 0; i < reply.n_elements(); i++) {
+        // Check that the ith entry is a string (i.e. key)
+        if (reply[i].redis_reply_type() != "REDIS_REPLY_STRING") {
+            throw SRRuntimeException("Element " + std::to_string(i) +
+                                     " in the aggregation list has an "\
+                                     "unexpected type: " +
+                                     reply.redis_reply_type());
+        }
+
+        // Check that the string length is not zero
+        if(reply[i].str_len() == 0) {
+            throw SRRuntimeException("Element " + std::to_string(i) +
+                                     " contains an empty key, which is "\
+                                     "not permitted.");
+        }
+
+        // Get the dataset key from the list entry
+        std::string dataset_key_prefix(reply[i].str(), reply[i].str_len());
+
+        // Build the metadata retrieval command
+        SingleKeyCommand* metadata_cmd =
+            metadata_cmd_list.add_command<SingleKeyCommand>();
+        (*metadata_cmd) << "HGETALL" << Keyfield(dataset_key_prefix + ".meta");
```

Comment:
Can we make a _build_metadata_key() routine so we don't have ".meta" embedded into functions like this? It'll make maintenance a lot easier later.

Reply:
I started doing this, but the ``_build`` routines we use for keys are actually incompatible with the aggregation list, and it requires a lot of code changes merge the two key construction procedures.  I made a note when doing that work to create an issue to rework how we do the key construction routines to make it easier.

Reply:
I'm good with that approach.  We definitely need some structure around key generation. I think it could be a good first issue for someone